Rewrite file hyperlinks on web

jennybc opened the pull request:
Addresses #7256

File hyperlinks in the Console should now open the targetted file, in a web/workbench context.

### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- N/A

#### Bug Fixes

- File hyperlinks in the R Console now open the relevant remote file when running on Workbench.

### QA Notes

@:web @:win @:console

The test that is being un-skipped here is a good manual test case. In an instance of Positron (Web), execute something like this:

``` r
library(cli)
cli_text("Let's open {.file DESCRIPTION:15:8}")
```

where `DESCRIPTION` is a path to a file that exists and the `:15` or `:15:8` part is an optional specification of line or line and column.

jennybc committed:
Unskip this test

jennybc committed:
Rewrite the file link for web

jennybc committed:
Remove annotation that explained the skip

jennybc committed:
Lean a bit more / a bit earlier into URI utilities

jennybc commented on src/vs/workbench/browser/positronAnsiRenderer/outputRun.tsx:
```diff
@@ -53,14 +55,20 @@ export const OutputRun = (props: OutputRunProps) => {
-		// Get the URL. If it's not a file URL, return it.
 		let url = props.outputRun.hyperlink.url;
-		if (!url.startsWith(`${Schemas.file}:`)) {
+		let uri: URI;
+		try {
+			uri = URI.parse(url);
+		} catch (e) {
+			console.error('Failed to parse URL:', e);
+			return url;
+		}
+
+		if (uri.scheme !== Schemas.file) {
```
Copilot proposed a more extensive refactoring of this function, leaning much more into working with the parsed URI and converting any line and column annotations into the form used internally. I declined to accepting much of this, in the spirit of "if it ain't broke, don't fix it".

But it does seem sensible to parse earlier and take easy opportunities to work with URI parts instead of treating the URL as a string and using regular expressions.

jennybc commented on src/vs/workbench/browser/positronAnsiRenderer/outputRun.tsx:
```diff
@@ -75,6 +83,19 @@ export const OutputRun = (props: OutputRunProps) => {
+		if (platform.isWeb) {
+			uri = uri.with({
+				scheme: Schemas.vscodeRemote,
+				authority: DOM.getActiveWindow().location.host
```
In the R sync, there was some question whether to use `DOM.getWindow(element)`, but there is no obvious `element` for me to target here and `DOM.getActiveWindow()` really does seem to be the most favorable thing to do.

Likewise, `uriToBrowserUri()` was also mentioned, but it doesn't preserve `path`, `query`, and `fragment`, all of which I want to leave undisturbed in this rewriting.

sharon-wang replied:
What do you think about constructing the URI by leveraging `toLocalResource`? It would involve passing the environmentService/pathService around as props in a bunch of places, so that we have access here, but then we can rely on our services to set the scheme and authority instead of using the DOM.

```ts
			uri = toLocalResource(
				uri,
				props.environmentService.remoteAuthority,
				props.pathService.defaultUriScheme
			);
```

jennybc commented:
I don't think the Windows test failure is related to this PR (?).

testlabauto commented:
> I don't think the Windows test failure is related to this PR (?).

Nope, it looks like there was a quarto install failure.

testlabauto commented on test/e2e/tests/console/console-r-hyperlinks.test.ts:
```diff
@@ -116,9 +116,7 @@ test.describe('Console Pane: R Hyperlinks', {
-	test.skip('R - Verify file hyperlink', {
-		annotation: [{ type: 'issue', description: 'https://github.com/posit-dev/positron/issues/7256' }],
-	}, async function ({ app, r }) {
+	test('R - Verify file hyperlink', async function ({ app, r }) {
```
Yes, I think we can probably test that.  The current line should have a different CSS class for the highlighting so that would at least cover the '15' part.  I would have to dig in more to see how we might detect the '8'.  I have made a note to add this test!

jennybc replied:
I think it would be fine to just test the line annotation (vs. line and column). I'll merge this PR once @sharon-wang signs off and, if we gain this other test in the future, that would be great.

testlabauto replied:
Sounds good.  Its on my list to add!

sharon-wang reviewed (dismissed):
Works well for me on Mac in Server Web and Desktop!

I have one suggestion for constructing the web URI -- let me know what you think! Otherwise, LGTM 👍

sharon-wang committed:
construct web uri with `toLocalResource`

this required updating various components to include `environmentService` and `pathService` in props/state, so that they can be used for constructing the web uri with `toLocalResource`.

jennybc commented on src/vs/workbench/browser/positronAnsiRenderer/outputRun.tsx:
```diff
@@ -30,11 +34,13 @@ const fileURLWithLineAndColumn = /^(file:\/\/\/.+):(\d+):(\d+)$/;
 
 /**
- * ColorType enumeration.
+ * ColorType enumeration.defaultUriAuthority
```
@sharon-wang Is this intentional?

sharon-wang replied:
nope! Thanks for catching this 😆

```suggestion
 * ColorType enumeration.
```

sharon-wang committed:
Remove unintentional lingering text

Signed-off-by: sharon <sharon-wang@users.noreply.github.com>
